From 13988c88d4b4a6f682a1555b98f8b60b89e22efd Mon Sep 17 00:00:00 2001
From: Flex1911 <dedsa2002@gmail.com>
Date: Fri, 7 Apr 2017 12:56:44 +0300
Subject: [PATCH] build: allow to disable recovery-two-step.img generation

Android 7.1.2 introduced new mechanism for recovery updating in two-step OTAs.
Although, recovery-two-step.img generation may be completely broken on some devices with custom mkbootimg handling.
We can use recovery.img as base for that devices, so recovery-two-step.img generation will not be required in this case.
Let's add new TARGET_NO_TWO_STEP_RECOVERY flag to skip recovery-two-step.img generation during target files packaging if we need it.

Change-Id: Icbea2da2f9565277622746545cdb96bf7f5ef3df
Signed-off-by: sub77 <sub77@ymail.com>
---
 core/Makefile                                 |  3 +++
 tools/releasetools/add_img_to_target_files.py | 16 +++++++++-------
 2 files changed, 12 insertions(+), 7 deletions(-)

diff --git a/core/Makefile b/core/Makefile
index 58d9ee9a2..ec62594a9 100644
--- a/core/Makefile
+++ b/core/Makefile
@@ -1990,6 +1990,9 @@ endif
 ifeq ($(BOARD_USES_FULL_RECOVERY_IMAGE),true)
 	$(hide) echo "full_recovery_image=true" >> $(zip_root)/META/misc_info.txt
 endif
+ifeq ($(TARGET_NO_TWO_STEP_RECOVERY),true)
+	$(hide) echo "no_two_step_recovery=true" >> $(zip_root)/META/misc_info.txt
+endif
 	$(call generate-userimage-prop-dictionary, $(zip_root)/META/misc_info.txt)
 #ifneq ($(INSTALLED_RECOVERYIMAGE_TARGET),)
 #	$(hide) PATH=$(foreach p,$(INTERNAL_USERIMAGES_BINARY_PATHS),$(p):)$$PATH MKBOOTIMG=$(MKBOOTIMG) \
diff --git a/tools/releasetools/add_img_to_target_files.py b/tools/releasetools/add_img_to_target_files.py
index 5a0a4116f..d8ac311d0 100755
--- a/tools/releasetools/add_img_to_target_files.py
+++ b/tools/releasetools/add_img_to_target_files.py
@@ -337,6 +337,7 @@ def AddImagesToTargetFiles(filename):
                                compression=zipfile.ZIP_DEFLATED)
 
   has_recovery = (OPTIONS.info_dict.get("no_recovery") != "true")
+  use_two_step_recovery = (OPTIONS.info_dict.get("no_two_step_recovery") != "true")
 
   def banner(s):
     print "\n\n++++ " + s + " ++++\n\n"
@@ -371,13 +372,14 @@ def AddImagesToTargetFiles(filename):
       if recovery_image:
         recovery_image.AddToZip(output_zip)
 
-      banner("recovery (two-step image)")
-      # The special recovery.img for two-step package use.
-      recovery_two_step_image = common.GetBootableImage(
-          "IMAGES/recovery-two-step.img", "recovery-two-step.img",
-          OPTIONS.input_tmp, "RECOVERY", two_step_image=True)
-      if recovery_two_step_image:
-        recovery_two_step_image.AddToZip(output_zip)
+      if use_two_step_recovery:
+        banner("recovery (two-step image)")
+        # The special recovery.img for two-step package use.
+        recovery_two_step_image = common.GetBootableImage(
+            "IMAGES/recovery-two-step.img", "recovery-two-step.img",
+            OPTIONS.input_tmp, "RECOVERY", two_step_image=True)
+        if recovery_two_step_image:
+          recovery_two_step_image.AddToZip(output_zip)
 
   banner("system")
   system_imgname = AddSystem(output_zip, recovery_img=recovery_image,
-- 
2.13.0

